web/src/message_list.js

```diff
@@ -340,8 +340,8 @@ export class MessageList {
         let just_unsubscribed = false;
         const subscribed = stream_data.is_subscribed_by_name(stream_name);
         const sub = stream_data.get_sub(stream_name);
-        const invite_only = sub.invite_only;
-        const is_web_public = sub.is_web_public;
+        const invite_only = sub && sub.invite_only;
+        const is_web_public = sub && sub.is_web_public;
```

Comment:
This probably makes the bug go away but I think it's extending a basically incoherent strategy in this function.

I think the better fix is to rewrite this using `narrow_state.stream_sub()` at the top, so that the `stream_name === undefined` check can instead be a `sub === undefined` check, and a lot of these additional lookups like `stream_data.is_subscribed_by_name(stream_name)` can just be checking properties of that `sub` object, and stop trying to work with the stream name as a source of information when we could be using an object.

Reply:
I didn't do that since there is some section of code below this, where it verifies if the `stream_sub` is undefined or not (indicating stream is deleted), and then marks it as deactivated. I guess this would be useful for rendering the stream grey, when you look up for it in search bar after deleting before reloading?

Reply:
@timabbott I guess that might not be a good idea -- currently, if the `stream_name` is undefined, then we return it, and in case of deleted streams, we don't want them to return since they would rather display ` This stream has been archived. `.

However, if we replace `stream_name` with `stream_sub`, then for deleted streams, `render_trailing_bookend` won't be executed with `deactivated=true`, and hence the ` This stream has been archived. ` message won't be displayed.

Reply:
This is in case when a deleted stream is searched for (without reload):<br>
<img width="555" alt="image" src="https://github.com/zulip/zulip/assets/97049524/3ecc2603-0635-4559-83ba-514b7a533c6c">
<br><br>
This is in case when a random stream is looked for:
<img width="563" alt="image" src="https://github.com/zulip/zulip/assets/97049524/f292af5f-057e-4882-bd58-6a4695137a7c">
<br><br>
This is in case when `stream_name` is replaced with `stream_sub`:
<img width="562" alt="image" src="https://github.com/zulip/zulip/assets/97049524/b448ed17-920e-4829-96ec-99b8785e5b18">

Reply:
OK, we should certainly avoid that symptom, but I don't see any difference at the implementation level between `narrow_state.stream_sub()` and the alternative that would cause that outcome -- they seem like they do the same logic.

Reply:
So, the difference is -- if you wanna replace [`stream_name`](https://github.com/zulip/zulip/blob/d99238526d8a0fd3fd244890a26f8c5978860522/web/src/message_list.js#L334) with `narrow_state.stream_sub()`, then you would be getting an early return at [`stream_sub === undefined`](https://github.com/zulip/zulip/blob/d99238526d8a0fd3fd244890a26f8c5978860522/web/src/message_list.js#L335) for deleted streams. This won't allow it to set the [`deactivated`](https://github.com/zulip/zulip/blob/d99238526d8a0fd3fd244890a26f8c5978860522/web/src/message_list.js#L348) flag as `true` for this stream, and render the `This stream has been archived message` from the [`render_trailing_bookend`](https://github.com/zulip/zulip/blob/d99238526d8a0fd3fd244890a26f8c5978860522/web/src/message_list.js#L353) method for deleted streams.

Reply:
Ah, OK, I had misread the `stream_name` function as behaving the same as `stream_sub` in terms of whether it's undefined.

Sure, given that, this is probably a reasonable path.